PackageReferenceProject.GetInstalledPackagesAsync API now only fetches installed packages

dominoFire opened the pull request:
<!-- DO NOT MODIFY OR DELETE THIS TEMPLATE. IT IS USED IN AUTOMATION. -->
## Bug

<!-- Search https://github.com/NuGet/Home/issues, and create one if you can't find a suitable issue. -->
<!-- Paste the full link, like https://github.com/nuget/home/issues/1000. GitHub will render is neatly. -->
Fixes: https://github.com/NuGet/Client.Engineering/issues/1780

Regression? Pes. Perf Regression Last working version: 6.2

## Description
<!-- Add details about the fix. Include any information that would help the maintainer review this change effective. -->

- Copied some logic from GetInstalledAndTransitivePackagesAsync (gitpa) API to only get installed packages. GetInstalledPackagesAsync (gipa) API only fetches installed packages.
- Each API, gitpa and gipa, has their own copy of cached Installed packages, to mitigate deadlocks. 
- Added unit tests
- `INuGetProject.GetInstalledPackagesAsync()` API is not affected, since it calls gitpa API

## PR Checklist

- [ ] PR has a meaningful title
- [ ] PR has a linked issue.
- [ ] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  <!-- Describe why you haven't added automation. -->
  - [ ] Test exception
  - **OR**
  - [ ] N/A <!-- Infrastructure, documentation etc. -->

- **Documentation**
  <!-- Please link the PR/issue if appropriate -->
  - [ ] Documentation PR or issue filled
  - **OR**
  - [ ] N/A


## Validation

- I validated code by running it on Experimental instance for 5 minutes on NuGet repo. No major crashes found.
- Vendors validated branch build 6.5.0.341, no new issues found.

Validated that IVsTestingExtension ran successfully with default options. INuGetProjectService.GetInstalledPackagesAsync API returns both installed and transitive packages.

When calling with Free Threaded check, it deadlocks. It also deadlocks before this fix.

![image](https://user-images.githubusercontent.com/1192347/203468964-962c7930-c3b2-47ee-9e7e-e4c7a6bb9971.png)

Validated with BenchmarkDotNet, it has a sightly better performance.


BenchmarkDotNet=v0.10.14, OS=Windows 10.0.22621, VM=Hyper-V
Intel Xeon Platinum 8370C CPU 2.80GHz, 1 CPU, 8 logical and 4 physical cores
  [Host]     : .NET Framework 4.6.1 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.8.9105.0
  DefaultJob : .NET Framework 4.6.1 (CLR 4.0.30319.42000), 64bit RyuJIT-v4.8.9105.0


Method |     Mean |     Error |    StdDev |      P50 |      P67 |      P95 |     P100 | Scaled |  Gen 0 | Allocated |
--------------------|---------:|----------:|----------:|---------:|---------:|---------:|---------:|-------:|-------:|----------:|
'GIPA Baseline' | 63.77 us | 0.5458 us | 0.5106 us | 63.89 us | 63.99 us | 64.40 us | 64.79 us |   1.00 | 1.2207 |   7.77 KB |
~~'GIPA Current PR (old)'~~ | 62.66 us | 0.3003 us | 0.2809 us | 62.74 us | 62.84 us | 62.97 us | 63.06 us |   0.98 | 0.9766 |   6.38 KB |
'GIPA Parametrized (current)' | 61.38 us | 0.7228 us | 0.6761 us | 61.24 us | 61.40 us | 62.39 us | 63.14 us |   0.96 | 0.7324 |   5.25 KB |

Commit:
First draft of GIPA api without transitive dependencies computation

Commit:
Added cancellation tokens

Commit:
Added unit tests

zivkan commented:
Please also use https://github.com/NuGet/Entropy/tree/main/IVsTestingExtension to test `INuGetProjectService.GetInstalledPackagesAsync` to make sure it still returns transitive packages.  Other VS components depend on this behaviour (one day I need to figure out how to write a reliable Apex test to validate this).

ghost commented:
This PR has been automatically marked as stale because it has no activity for **7 days**. It will be closed if no further activity occurs **within another 7 days of this comment**. If it is closed, you may reopen it anytime when you're ready again, as long as you don't delete the branch.

Commit:
GIPAEventSource logging via ETW

Added GIPAEventSource for GIPA API ETL logging

Increased resolution on duration to ticks

Fixed error in GIPAEventSource LogCall

Commit:
Removed GITPA and GIPA shared locks and cache

Commit:
Fixed bug in caching InstalledPackages

Commit:
Revert "GIPAEventSource logging via ETW"

This reverts commit 6248db48a24c2f0c05761bf889d5963f85dc8d48.

Commit:
Tests: Added tests for GetInstalledPackagesInternalAsync

Commit:
Revert "Tests: Added tests for GetInstalledPackagesInternalAsync"

This reverts commit 65b23b173c1448633d9f8f0f93c56488f396d205.

dominoFire commented:
This PR was affected by a unstable test

https://github.com/NuGet/Home/issues/9476

##[warning]RunMessage : [xUnit.net 00:15:02.72] NuGet.Tests.Apex: [Long Running Test] 'NuGet.Tests.Apex.AuthorSignedPackageTestCase.InstallFromPMCForPC_SucceedAsync', Elapsed: 00:05:23

erdembayar commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:
```diff
@@ -91,12 +95,61 @@ public override async Task<IReadOnlyList<PackageSpec>> GetPackageSpecsAsync(Depe
         {
-            ProjectPackages packages = await GetInstalledAndTransitivePackagesAsync(includeTransitiveOrigins: false, token);
-            return packages.InstalledPackages;
+            token.ThrowIfCancellationRequested();
```
Could this have implemented as conditional in `GetInstalledAndTransitivePackagesAsync`?
As current state it's not really large method, I'm not really worried about, but still code duplication.

dominoFire replied:
I tried simplifying to one method here https://github.com/NuGet/NuGet.Client/commit/e7d351a2aa230fb304694919627fdab1d7b5ad06

but it has worse performance results if I have only one method. See [comment](https://github.com/NuGet/Client.Engineering/issues/1780#issuecomment-1324299876)

dominoFire replied:
Chatted offline, will try this approach by parametrizing `GetInstalledAndTransitivePackagesAsync` and profiling with benchmarkdotnet inspired from [zivkan/Benchmarks](https://github.com/zivkan/Benchmarks) repo sample

dominoFire replied:
In 2 benchmark runs, the parametrized approach performs better than the approach in this current PR (duplicated code). [Benchmark code available in another branch](https://github.com/NuGet/NuGet.Client/compare/dev...dev-dominofire-gipa-benchmark)


Method |     Mean |     Error |    StdDev |      P50 |      P67 |      P95 |     P100 | Scaled | ScaledSD |  Gen 0 | Allocated |
--------------------|---------:|----------:|----------:|---------:|---------:|---------:|---------:|-------:|---------:|-------:|----------:|
'GIPA Baseline' | 63.94 us | 0.7366 us | 0.6890 us | 63.91 us | 64.19 us | 64.88 us | 65.28 us |   1.00 |     0.00 | 1.2207 |   7.77 KB |
'GIPA Current PR' | 62.88 us | 0.8915 us | 0.8339 us | 62.63 us | 63.14 us | 64.33 us | 64.60 us |   0.98 |     0.02 | 0.9766 |   6.38 KB |
'GIPA Parametrized' | 60.95 us | 0.3463 us | 0.2892 us | 60.91 us | 61.03 us | 61.38 us | 61.67 us |   0.95 |     0.01 | 0.7935 |   5.25 KB |




Method |     Mean |     Error |    StdDev |      P50 |      P67 |      P95 |     P100 | Scaled | ScaledSD |  Gen 0 | Allocated |
--------------------|---------:|----------:|----------:|---------:|---------:|---------:|---------:|-------:|---------:|-------:|----------:|
'GIPA Baseline' | 64.25 us | 0.8349 us | 0.7810 us | 64.39 us | 64.80 us | 65.16 us | 65.28 us |   1.00 |     0.00 | 1.2207 |   7.77 KB |'GIPA Current PR' | 62.41 us | 0.5463 us | 0.5110 us | 62.53 us | 62.62 us | 63.14 us | 63.27 us |   0.97 |     0.01 | 0.9766 |   6.38 KB |
'GIPA Parametrized' | 61.28 us | 0.9516 us | 0.8436 us | 61.27 us | 61.36 us | 62.95 us | 63.00 us |   0.95 |     0.02 | 0.7324 |   5.25 KB |

erdembayar reviewed (commented):
Just 1 comment. If you feel strongly toward current implementation, then it's still ok.

kartheekp-ms commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:
```diff
@@ -91,12 +95,61 @@ public override async Task<IReadOnlyList<PackageSpec>> GetPackageSpecsAsync(Depe
+            if (IsInstalledComputationNeeded)
+            {
+                installedPackages = new T();
+                targetsList = (await GetTargetsListAsync(assetsFilePath, token))?.ToList();
```
Just curious, do we need `.ToList();` here?

dominoFire replied:
Good catch. Another API requires IReadOnlyList, I've thining on chaning that API to avoid this call.

dominoFire replied:
Removed all `.ToList()` calls in `GetTargetsListAsync()`

kartheekp-ms commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:
```diff
@@ -91,12 +95,61 @@ public override async Task<IReadOnlyList<PackageSpec>> GetPackageSpecsAsync(Depe
+                .SelectMany(f => ResolvedInstalledPackagesList(f.Dependencies, f.FrameworkName, targetsList, installedPackages))
+                .GroupBy(p => p.PackageIdentity)
+                .Select(g => g.OrderBy(p => p.TargetFramework, FrameworkSorter).First())
+                .ToList();
```
Just curious, do we need .ToList(); here?

nkolev92 replied:
Despite the fact that the API returns IEnumerable, it is preferable to return a fully evaluated collection. 
If you don't do that and the types change you can get an exception when you use the returned collection.

dominoFire replied:
To materialize results before sending back.

Commit:
Changed targetList APIs to avoid ToList() calls

Commit:
Tests: Fixed tests

Commit:
Parametrized GITPA to use in GIPA API

Commit:
nit

zivkan commented:
@dominoFire did you verify the `INuGetProjectService.GetInstalledPackagesAsync()` API, that we provide to other VS, still returns transitive packages? https://github.com/NuGet/NuGet.Client/pull/4900#issuecomment-1307797575

I hoped to validate myself, but just don't have time, and the title of this PR makes it sound like this PR would break it (it's not quickly obvious to me looking at the changes whether it would or wouldn't).

zivkan commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:
```diff
@@ -150,83 +185,101 @@ public async Task<ProjectPackages> GetInstalledAndTransitivePackagesAsync(bool i
+        protected abstract IEnumerable<PackageReference> ResolvedInstalledPackagesList(IEnumerable<LibraryDependency> libraries, NuGetFramework targetFramework, IList<LockFileTarget> targets, T installedPackages);
 
-        protected abstract IReadOnlyList<PackageReference> ResolvedTransitivePackagesList(NuGetFramework targetFramework, IReadOnlyList<LockFileTarget> targets, T installedPackages, T transitivePackages);
+        protected abstract IReadOnlyList<PackageReference> ResolvedTransitivePackagesList(NuGetFramework targetFramework, IList<LockFileTarget> targets, T installedPackages, T transitivePackages);
```
I don't understand why the `targets` type was changed from `IReadOnly<T>` to `<IList<T>`. I don't see where it's being mutated.

Haven't we already had an instance of `PackageReferenceProject.GetInstalledPackageAsync()` not being thread-safe?  Does this change add risk that the implementation might become non-thread safe again?

dominoFire replied:
See changes in PackageReferenceProject.cs

![image](https://user-images.githubusercontent.com/1192347/204672776-d2c51a52-4ca1-4e30-8427-0d6ea81c4fde.png)

IList has methods to mutate the list, whereas `IReadOnly...<T>` not, and I want to avoid type conversions to avoid performance issues.

dominoFire replied:
Regarding thread-safety, this PR calls GetInstalledAndTransitivePackageAsync(), which is a method that is considered thread-safe. 

There are tests that fall into this scenario, see 

https://github.com/NuGet/NuGet.Client/blob/c0371f7165bea201aded1a8013ca0b17e2ad0092/test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/CpsPackageReferenceProjectTests.cs#L3908


I can add a custom test for this API, if you want.

dominoFire replied:
Also, there's a test in this PR to test GetInstalledPackagesAsync With GetInstalledAndTransitivePackagesAsync calls at the same time.

GetInstalledPackagesAsync_WithGetInstalledAndTransitivePackagesCallAtSameTime_DoesNotDeadlockAsync

dominoFire replied:
Added new test: GetInstalledPackagesAsync_WithMultipleCalls_DoesNotDeadlockAsync

dominoFire commented:
> @dominoFire did you verify the `INuGetProjectService.GetInstalledPackagesAsync()` API, that we provide to other VS, still returns transitive packages? [#4900 (comment)](https://github.com/NuGet/NuGet.Client/pull/4900#issuecomment-1307797575)
> 
> I hoped to validate myself, but just don't have time, and the title of this PR makes it sound like this PR would break it (it's not quickly obvious to me looking at the changes whether it would or wouldn't).

@zivkan  I validated it using the IVsTestingExtension Entropy tool. See PR description. Also, I saw that `INuGetProjectService.GetInstalledPackagesAsync()` is calling `PackageReferenceProject.GetInstalledAndTransitivePackagesAsync()`  https://github.com/NuGet/NuGet.Client/blob/c0371f7165bea201aded1a8013ca0b17e2ad0092/src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/NuGetProjectService.cs#L163

Commit:
Tests: Added test

erdembayar commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs:
```diff
@@ -107,36 +113,65 @@ public async Task<ProjectPackages> GetInstalledAndTransitivePackagesAsync(bool i
+                        // Make a copy of the caches to prevent concurrency issues.
+                        lock (_installedAndTransitivePackagesLock)
+                        {
+                            transitivePackages = GetCollectionCopy(TransitivePackages);
```
Where is the actual new calculation is happening for `transitivePackages`? I didn't visit this code path since last code review. Is it happening inside `ResolvedTransitivePackagesList`?

dominoFire replied:
Yes, that's correct.
